integration/run_test.go

```diff
@@ -185,3 +191,216 @@ func testWebServiceContainer(shell *shell.Shell, containerName string) {
+				ExpectedSuccess:    false,
+			},
+		},
+		{
+			name: "2s network interruption, ~3-4s retries allowed, success",
+			config: retryConfig{
+				MaxRetries:         5,
+				MinWaitMsec:        100,
+				MaxWaitMsec:        1600,
+				NetworkDisableMsec: 2000,
+				ExpectedSuccess:    true,
+			},
+		},
+		{
+			name: "5s network interruption, ~1-2s retries allowed, failure",
+			config: retryConfig{
+				MaxRetries:         1,
+				MinWaitMsec:        100,
+				MaxWaitMsec:        1600,
+				NetworkDisableMsec: 5000,
+				ExpectedSuccess:    false,
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			regConfig := newRegistryConfig()
+			sh, done := newShellWithRegistry(t, regConfig)
+			defer done()
+
+			registryHostIP := getIP(t, sh, regConfig.host)
+
+			config := `
+[blob]
+max_retries   = ` + strconv.Itoa(tt.config.MaxRetries) + `
+min_wait_msec = ` + strconv.FormatInt(tt.config.MinWaitMsec, 10) + `
+max_wait_msec = ` + strconv.FormatInt(tt.config.MaxWaitMsec, 10) + `
+
+[background_fetch]
+disable = true
+`
+
+			m := rebootContainerd(t, sh, getContainerdConfigToml(t, false), getSnapshotterConfigToml(t, false, config))
+			// Mirror image
+			copyImage(sh, dockerhub(containerImage), regConfig.mirror(containerImage))
+			// Pull image, create SOCI index with all layers and small (100kiB) spans
+			indexDigest := buildIndex(sh, regConfig.mirror(containerImage), withMinLayerSize(0), withSpanSize(100*1024))
+			sh.X("soci", "image", "rpull", "--user", regConfig.creds(), "--soci-index-digest", indexDigest, regConfig.mirror(containerImage).ref)
+
+			// Run the container
+			image := regConfig.mirror(containerImage).ref
+			sh.X("soci", "run", "-d", "--snapshotter=soci", image, "test-container")
+
+			sh.X("apt-get", "--no-install-recommends", "install", "-y", "iptables")
+
+			// TODO: Wait for the container to be up and running
+
+			// Block network access to the registry
+			if tt.config.NetworkDisableMsec > 0 {
+				sh.X("iptables", "-A", "OUTPUT", "-d", registryHostIP, "-j", "DROP")
+			}
+
+			// Do something in the container that should work without network access
+			commandSucceedStdout, _, err := sh.R("ctr", "task", "exec", "--exec-id", "test-task-1", "test-container", "sh", "-c", "times")
+			if err != nil {
+				t.Fatalf("attempt to run task without network access failed: %s", err)
+			}
+
+			type ErrorLogLine struct {
+				Error string `json:"error"`
+				Msg   string `json:"msg"`
+			}
+
+			gaveUpChannel := make(chan bool, 1)
+			defer close(gaveUpChannel)
+
+			monitorGaveUp := func(rawL string) {
+				if i := strings.Index(rawL, "{"); i > 0 {
+					rawL = rawL[i:] // trim garbage chars; expects "{...}"-styled JSON log
+				}
+				var logLine ErrorLogLine
+				if err := json.Unmarshal([]byte(rawL), &logLine); err == nil {
+					if logLine.Msg == "statFile error" && strings.Contains(logLine.Error, "giving up after") {
+						gaveUpChannel <- true
+						return
+					}
+				}
+			}
```

Comment:
this func seems quite similar to 

https://github.com/awslabs/soci-snapshotter/blob/1c621e874ef015932b0d9a6db0053d2484ad4278/util/testutil/shell.go#L206

Reply:
It is. Unfortunately not close enough that I could conveniently combine them without something like the refactor described in https://github.com/awslabs/soci-snapshotter/pull/499#discussion_r1165895996 which I do intend to follow up on at a lower priority.

Reply:
Yeah maybe we an seperate out this logic into a wrapper

```go
func monitorWrapper(monitorFunc func(string)) {
   if i := strings.Index(rawL, "{"); i > 0 {
        rawL = rawL[i:] // trim garbage chars; expects "{...}"-styled JSON log
    }
   var logLine ErrorLogLine
   if err := json.Unmarshal([]byte(rawL), &logLine); err == nil {
     monitorFunc()
   }
}

func (m *LogMonitor) Add(name string, monitorFunc func(string)) error {
	if _, ok := m.monitorFuncs[name]; ok {
		return fmt.Errorf("attempted to add log monitor with already existing name: %s", name)
	}
	m.monitorFuncs[name] = monitorWrapper(monitorFunc)
	return nil
}

}

```

Reply:
The `ErrorLogLine` type in your suggestion is different in the two use cases.

Reply:
Oh sorry yeah I meant once we have a generalized type to unmarshal to, like you were referring to above

Reply:
> It is.

Then we're having 3 functions with different structures for the similar purpose:

1. here, a local func variable;
2. [`monitorStartup`](https://github.com/awslabs/soci-snapshotter/blob/1c621e874ef015932b0d9a6db0053d2484ad4278/util/testutil/shell.go#L206): a standalone func;
3. [`MonitorFunc`](https://github.com/awslabs/soci-snapshotter/blob/1c621e874ef015932b0d9a6db0053d2484ad4278/util/testutil/shell.go#L154): a struct method.

Any thoughts on which forces us to have such a structure.

Reply:
At the moment, expedience and wanting to delay generalized implementation until we have more use cases. Now that we have three, I have a refactor in mind that would combine at least two of them, possibly all three, but it would touch all three so I don't want to put it in a PR implementing just one of them.

Reply:
https://github.com/awslabs/soci-snapshotter/issues/567